### PR TITLE
Added deep_transform_keys to StrongParameters

### PR DESCRIPTION
**Pull request by gustavogr**
### Summary

This PR implements the methods `deep_transform_keys` and `deep_transform_keys!` for `ActionController::Parameters`.

### Other Information

Thread in the Ruby On Rails Mailing list can be found [here](https://groups.google.com/forum/#!topic/rubyonrails-core/paiccJIC5E0).

### CODE REVIEW

File: actionpack/lib/action_controller/metal/strong_parameters.rb
Comment (eugeneius) [inline actionpack/lib/action_controller/metal/strong_parameters.rb]: It looks like this was copied from [the `transform_keys` implementation](https://github.com/rails/rails/blob/aa7da0471f7fcf2adbbc26c37a25914d8493928d/actionpack/lib/action_controller/metal/strong_parameters.rb#L687). It makes sense there, because calling `transform_keys` on a hash without passing a block returns an enumerator:

```ruby
irb(main):001:0> { foo: "bar" }.transform_keys
=> #<Enumerator: {:foo=>"bar"}:transform_keys>
```

However that doesn't work for `deep_transform_keys`:

```ruby
irb(main):002:0> { foo: "bar" }.deep_transform_keys
Traceback (most recent call last):
        7: from tools/console:11:in `<main>'
        6: from (irb):2
        5: from /Users/eugene/src/rails/activesupport/lib/active_support/core_ext/hash/keys.rb:66:in `deep_transform_keys'
        4: from /Users/eugene/src/rails/activesupport/lib/active_support/core_ext/hash/keys.rb:119:in `_deep_transform_keys_in_object'
        3: from /Users/eugene/src/rails/activesupport/lib/active_support/core_ext/hash/keys.rb:119:in `each_with_object'
        2: from /Users/eugene/src/rails/activesupport/lib/active_support/core_ext/hash/keys.rb:119:in `each'
        1: from /Users/eugene/src/rails/activesupport/lib/active_support/core_ext/hash/keys.rb:120:in `block in _deep_transform_keys_in_object'
LocalJumpError (no block given (yield))
```

Which means this method's implementation could be simplified:

```ruby
    def deep_transform_keys(&block)
      new_instance_with_inherited_permitted_status(
        @parameters.deep_transform_keys(&block)
      )
    end
```

Comment (GustavoTango) [reply actionpack/lib/action_controller/metal/strong_parameters.rb]: Thank you for the feedback. I pushed a commit with your suggestion.

Comment (gustavogr) [reply actionpack/lib/action_controller/metal/strong_parameters.rb]: Thank you for your feedback. Pushed a commit with your suggestion.

**Comment by gustavogr** [comment]
Added a couple of tests in `accessors_test.rb` after looking at the changes in #36302. I rebased and squashed the changes in one commit following the contributors guide.
This is my first time contributing to Rails, so if there is more work to be done I'm happy to get to it.

File: actionpack/CHANGELOG.md
Comment (eugeneius) [inline actionpack/CHANGELOG.md]: StrongParameters looks like a module name here. Could you either change it to "strong parameters", or the actual class name (`ActionController::Parameters`)?

File: actionpack/lib/action_controller/metal/strong_parameters.rb
Comment (eugeneius) [inline actionpack/lib/action_controller/metal/strong_parameters.rb]: The documentation for `deep_transform_keys` mentions nested hashes **and arrays**:

https://github.com/rails/rails/blob/bbd065204ed45071f048badb7c0d49e35773d788/activesupport/lib/active_support/core_ext/hash/keys.rb#L58-L59

Could you include that here too? Obviously arrays don't have keys, but the point is that the method traverses nested arrays of hashes, not just nested hashes. If you can think of a clearer way to explain that, all the better 😊

File: actionpack/lib/action_controller/metal/strong_parameters.rb
Comment (eugeneius) [inline actionpack/lib/action_controller/metal/strong_parameters.rb]: Ditto on mentioning nested arrays here.

File: actionpack/lib/action_controller/metal/strong_parameters.rb
Comment (gustavogr) [reply actionpack/lib/action_controller/metal/strong_parameters.rb]: Maybe something like "_This method includes the keys of nested hashes and array of hashes_"?

**Comment by gustavogr** [comment]
Pushed new commit with changes in the documentation. Let me know if the documentation is still not clear.

File: actionpack/lib/action_controller/metal/strong_parameters.rb
Comment (eugeneius) [inline actionpack/lib/action_controller/metal/strong_parameters.rb]: The grammatically correct phrasing here would be "arrays of hashes".

I think I prefer the version used in the main `deep_transform_keys` docs though: "nested hashes and arrays of hashes" implies to me that only one level of nesting is covered, and the keys of a hash within an array-of-arrays wouldn't be transformed, for example, when in fact they would.

**Comment by gustavogr** [comment]
Changed the documentation to match the version used in the main `deep_transform_keys` docs.

**Comment by eugeneius** [comment]
This now looks good to me, thanks @gustavogr!

A committer will have to take a look before it can be merged.

**Review by eugeneius — APPROVED** [review]

**Comment by gustavogr** [comment]
Thanks for your detailed help! @eugeneius

**Pull request merged**
